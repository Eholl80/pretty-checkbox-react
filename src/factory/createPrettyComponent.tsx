import * as React from 'react';
import classNames from 'classnames';

import { getClassNames, getChecked } from './utils';
import { CommonCheckboxRadioProps, SwitchShape } from '../types/CommonProps';

const createInput = ({
    onChange,
    disabled,
    value,
    state,
    locked,
    type,
    name,
    checked,
}: CommonCheckboxRadioProps) => {
    return (
        <input
            onChange={onChange}
            type={type}
            value={value}
            name={name}
            checked={
                typeof checked === 'undefined'
                    ? getChecked({ state, value, disabled, locked, type })
                    : checked
            }
            disabled={disabled}
        />
    );
};

const createLabel = ({ color, icon, children }: CommonCheckboxRadioProps) => (
    <div
        className={classNames('state', {
            [`p-${color}`]: color,
        })}>
        {icon}
        <label>{children}</label>
    </div>
);

export const createPrettyComponent = (
    props: Omit<CommonCheckboxRadioProps, 'icon' | 'shape'> & {
        as?: string | React.FunctionComponent | React.ComponentClass;
        icon?: React.ReactElement;
        shape?: SwitchShape | CommonCheckboxRadioProps['shape'];
        isSwitch?: boolean;
    }
) => {
    const {
        as = 'div',
        iconType,
        type,
        children,
        animation,
        icon,
        bigger,
        plain,
        shape,
        fill,
        className,
        setState,
        state,
        locked,
        disabled,
        onChange,
        checked,
        value,
        color,
        name,
        ...rest
    } = props;

    return React.createElement<any>(
        as,
        {
            className: getClassNames(props),
            'aria-disabled': disabled,
            'aria-checked': state === 'indeterminate' ? 'mixed' : !!state,
            role: type,
            ...rest,
        },
        <>
            {createInput(props)}
            {createLabel(props)}
        </>
    );
};
