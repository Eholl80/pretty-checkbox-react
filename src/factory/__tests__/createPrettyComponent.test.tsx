import * as React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { createPrettyComponent } from '../createPrettyComponent';

const createComponent = (
    userProps: any = {}
): {
    container: HTMLElement;
    props: {
        onChange: jest.Mock;
        state: any;
    };
} => {
    const props = {
        state: false,
        onChange: jest.fn(),
        role: 'checkbox',
        type: 'checkbox',
        ...userProps,
    };

    const { container } = render(createPrettyComponent(props));

    return {
        container,
        props,
    };
};

describe('createPrettyComponent tests', () => {
    it('should render without errors', () => {
        expect(() => {
            createComponent().container;
        }).not.toThrow();
    });

    it('it should match the base snapshot', () => {
        expect(createComponent().container).toMatchSnapshot();
    });

    it.skip('should trigger onChange when input is changed', () => {
        const { container, props } = createComponent();
        const input = container.querySelector('input');

        if (input) {
            fireEvent.change(input, { target: { value: '' } });
        }

        expect(props.onChange).toHaveBeenCalledTimes(1);
    });
});
