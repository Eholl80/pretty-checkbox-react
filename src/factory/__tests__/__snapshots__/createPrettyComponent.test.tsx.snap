// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createPrettyComponent tests it should match the base snapshot 1`] = `
<div>
  <div
    aria-checked="false"
    class="pretty p-default"
    role="checkbox"
  >
    <input
      type="checkbox"
      value=""
    />
    <div
      class="state"
    >
      <label />
    </div>
  </div>
</div>
`;
